### DISCUSSION TITLE
React native dapp migration

### DISCUSSION THREAD

**Discussion by emzet93** [WalletConnect/walletconnect-monorepo#2556] [Developer Support]
I'm looking at all docs/tutorials about migration to v2 on react native dapps and it looks like the main supported way is to use `@web3modal/react-native`. It's like 3 weeks before v1 sunset and this sdk is still in alpha and doesn't seem stable enough to use in production?

Anyway, I played with `@web3modal/react-native` and it's really cool, but it goes with its own modal UI which is not customisable at all. It also requires external dependencies of `react-native-modal`, it uses `valtio` and it basically seems like whole framework/mini app. The thing I liked a lot with v1 `@walletconnect/react-native-dapp` is that it used just react context for state management and it allowed to be used fully headless, providing custom render props to whole modal content.

Is sth like that doable on v2 with `@web3modal/react-native`? If not, are there any headless sdks/libs or examples to build own UI for that?

Basically all I need to do is to fetch supported wallets list, get some methods to connect to provided wallet, and then get address of connected wallet and provider allowing signing messages. I want to provide custom UI for that and not sure where to look for resources. Maybe just source code of `@web3modal/react-native` is a good start?

**Comment by ignaciosantise** [comment]
hey @emzet93! Thanks for the feedback.

We don't provide a way to render a custom modal but we are already discussing this possibility, so we'll take your feedback to the discussion too. Can you give us more context about this request? Is it because you want to show less wallets? Or you want to add a UI that matches your app style?

You can find the source code here: https://github.com/WalletConnect/web3modal-react-native

**Reply by emzet93** [reply]
> hey @emzet93! Thanks for the feedback.
> 
> 
> 
> We don't provide a way to render a custom modal but we are already discussing this possibility, so we'll take your feedback to the discussion too. Can you give us more context about this request? Is it because you want to show less wallets? Or you want to add a UI that matches your app style?
> 
> 
> 
> You can find the source code here: https://github.com/WalletConnect/web3modal-react-native

Actually both. Having UI matching vibe of whole app is super important to me. But limiting/ordering available wallets too. Basically would be good to have more control over plug-in and having headless api would be great way to achieve that. Also having dependency of react-native-modal seems unnecessary if I don't use it in my project.

I'll check provided source code, thanks!

Btw, it's 2 weeks before the v1 sunset. Do you think that web3modal for react native will get stable before that time?

**Comment by felipecamposfabel** [comment]
I'm facing the same issue as @emzet93. We have a custom UI to match our brand and to give a better UX by listing and ordering installed wallets, and popular ones, and so on... I know we could use your provider + fetch wallets from your explorer API, but that's pretty much what you already do inside your Web3Modal component and Controllers.

Do you think it would be possible, before sunset ofc, to expose a hook that initializes the provider, like `useConfigure` does, but returns similar props to what is passed to the `renderQrcodeModal` function in V1? So we could work with `provider.connect()` or `uri` (or a provider listener on `diplay_uri`) + walletServices, without the need for an entire visual component.

Please, kindly, let me know if it's something you are considering so I can wait or I should do it myself right away. Thanks!

**Comment by ignaciosantise** [comment]
Hey @emzet93 👋 web3modal is stable right now, this week we'll create a release candidate version with the current features. And after the release, we'll add props to reorder/exclude wallets, and add custom wallets.

@felipecamposfabel unfortunately it won't be possible to expose a hook to use the provider separated from the UI yet

**Comment by lunajuan** [comment]
@felipecamposfabel and @emzet93 may I ask how you were able to get this to work? Did you have to basically recreate this package from scratch or were you able to patch it? For example, remove the UI part in favor of your own component within the package. It's not clear to me what the best path is. I feel like I'm missing something since the docs say the following regarding `useWalletConnectModal`:

https://docs.walletconnect.com/2.0/reactnative/walletconnectmodal/dapp-usage#usewalletconnectmodal
> Useful when you want to use your own UI elements and subscribe to modals state.

**Reply by felipecamposfabel** [reply]
@lunajuan We just shipped our App with wallet connect v2 using a custom hook. I just forked their repo and created a custom hook based on what the Modal already does. I have no idea why they didn't create a similar hook, used it on their modal, and exposed it...

The hook I created handles some events (that the modal also handles) and returns something like:

```
{
      connectToWalletService, // to connect to wallets
      isConnected, // if is connected
      address, // connected address
      provider, // eth provider
      uri, // uri (not using, but good for debugging)
      wallets, // list of compatible wallets for customized list
      connect, // start the connection with WC
}
```

**Reply by lunajuan** [reply]
Thank you @felipecamposfabel. I tried this and got my custom WC integration working again. Congratulations on shipping with WC v2!
